### PR TITLE
Standardize dependency management

### PR DESCRIPTION
**Pull request by satyaog**
- [x] BERT_pytorch
- [x] Background_Matting
- [x] LearningToPaint
- [x] Super_SloMo
- [x] alexnet
- [x] attention_is_all_you_need_pytorch
- [x] dcgan
- [x] demucs
- [x] densenet121
- [x] detectron2_maskrcnn
- [x] dlrm
- [x] drq
- [x] fastNLP_Bert
- [x] hf_Albert
- [x] hf_Bart
- [x] hf_Bert
- [x] hf_BigBird
- [x] hf_DistilBert
- [x] hf_GPT2
- [x] hf_Longformer
- [x] hf_Reformer
- [x] hf_T5
- [x] maml
- [x] maml_omniglot
- [x] mnasnet1_0
- [x] mobilenet_v2
- [x] mobilenet_v2_quantized_qat
- [x] mobilenet_v3_large
- [x] moco
- [x] nvidia_deeprecommender
- [x] opacus_cifar10
- [x] pyhpc_equation_of_state
- [x] pyhpc_isoneutral_mixing
- [x] pyhpc_turbulent_kinetic_energy
- [x] pytorch_CycleGAN_and_pix2pix
- [x] pytorch_stargan
- [x] pytorch_struct
- [x] pytorch_unet
- [x] resnet18
- [x] resnet50
- [x] resnet50_quantized_qat
- [x] resnext50_32x4d
- [x] shufflenet_v2_x1_0
- [x] soft_actor_critic
- [x] speech_transformer
- [?] squeezenet1_1
- [x] tacotron2
- [ ] timm_efficientdet
- [x] timm_efficientnet
- [x] timm_nfnet
- [x] timm_regnet
- [x] timm_resnest
- [x] timm_vision_transformer
- [x] timm_vovnet
- [x] tts_angular
- [x] vgg16
- [x] vision_maskrcnn
- [x] yolov3

### CODE REVIEW

**Comment by satyaog** [comment]
@breuleux , @abergeron , this is rather small and just to make sure that I'm not missing something that I should also do in the case of `torchvision`

**Comment by breuleux** [comment]
@satyaog I think that suffices for the torchvision suite. We can store the frozen reqs in `requirements.txt`, given that we have `requirements.in` for the loose specifications.

**Comment by satyaog** [comment]
Cool thanks. I was reproducing the filenames of `https://github.com/mila-iqia/milabench/tree/9c9fbf43a1389a1416330128f63a6ce511aa410f/benchmarks/pytorch-lightning` with `frozen_reqs.*.txt`. Would you like me to also rename the files of pytorch-lightning for consistency?

File: benchmarks/stable_baselines3/benchfile.py [hunk 1]
```diff
@@ -9,19 +9,14 @@ class StableBenchmarkPack(Package):
 
     def install(self):
         """Install requirements one by one, as baselines setup is non-standard"""
+        # box2d-py requires swig
+        self.conda_install("-c", "conda-forge", "swig")
```
Comment (breuleux) [inline benchmarks/stable_baselines3/benchfile.py:13]: Can this work? https://pypi.org/project/swig/

File: benchmarks/torchbench/benchfile.py [hunk 2]
```diff
@@ -54,19 +54,20 @@ def install(self):
...
-            # run the global install.py, so we pass resnet18 which has an empty
-            # install.py.
-            self.python("install.py", "--models", "resnet18")
-        else:
+            reqs = f"requirements-{model_name}"
+
+        if reqs:
+            self.pip_install("-r", code / f"{reqs}.txt")
+            if (code / f"{reqs}-extra.txt").exists():
+                self.pip_install("-r", code / f"{reqs}-extra.txt")
```
Comment (breuleux) [inline benchmarks/torchbench/benchfile.py:68]: What are these?

File: benchmarks/stable_baselines3/benchfile.py (hunk 1 above)
Comment (satyaog) [reply benchmarks/stable_baselines3/benchfile.py:13]: Interesting I'll check

File: benchmarks/torchbench/benchfile.py (hunk 2 above)
Comment (satyaog) [reply benchmarks/torchbench/benchfile.py:68]: Name is probably quite bad but this is specifically for `requirements-detectron2_maskrcnn-extra.txt` which contains `detectron2` for which the `setup.py` requires `torch` to be installed : https://github.com/facebookresearch/detectron2/blob/c470ca31d2b54d038b9e78f48d1d8e1a8debcc68/setup.py

**Comment by satyaog** [comment]
### LearningToPaint.0.json

is it normal to have a loss of -0.0?

```json
{"#start": 1676344299.2643087}
{"#config": {"dirs": {"code": "/network/scratch/o/ortizgas/public/bench_reqs/code/LearningToPaint", "venv": "/network/scratch/o/ortizgas/public/bench_reqs/venv/LearningToPaint", "data": "/network/scratch/o/ortizgas/public/bench_reqs/data", "runs": "/network/scratch/o/ortizgas/public/bench_reqs/runs", "base": "/network/scratch/o/ortizgas/public/bench_reqs"}, "venv": {"type": "virtualenv"}, "plan": {"method": "njobs", "n": 1}, "voir": {"--forward": ["#stdout", "#stderr", "#end", "loss", "reward", "compute_rate", "train_rate", "loading_rate", "descr", "progress", "total", "gpu_time", "cpu_dispatch_time", "cpu_walltime", "tflops"], "--compute-rate": true, "--train-rate": true, "--loading-rate": true}, "argv": {"-d": "cuda", "-m": "eager", "-t": "train"}, "config_base": "/home/satya-local/CODE/milabench_reqs/benchmarks/torchbench", "config_file": "/home/satya-local/CODE/milabench_reqs/benchmarks/torchbench/benchtest.yaml", "definition": "/home/satya-local/CODE/milabench_reqs/benchmarks/torchbench", "model": "LearningToPaint", "name": "LearningToPaint", "group": "LearningToPaint", "tag": ["LearningToPaint", "0"], "job-number": 0}}
{"#stdout": "Running train method from LearningToPaint on cuda in eager mode."}
{"#stdout": "\n"}
{"loss": 23239162.0}
{"#stderr": "/network/scratch/o/ortizgas/public/bench_reqs/code/LearningToPaint/torchbenchmark/models/LearningToPaint/baseline/DRL/ddpg.py:167: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).\n  s0 = torch.tensor(self.state, device='cpu')\n"}
{"#stderr": "/network/scratch/o/ortizgas/public/bench_reqs/code/LearningToPaint/torchbenchmark/models/LearningToPaint/baseline/DRL/ddpg.py:170: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).\n  s1 = torch.tensor(state, device='cpu')\n"}
{"loss": 1435457.125}
{"loss": -0.0}
{"loss": -0.0}
{"loss": -0.0}
{"loss": -0.0}
{"loss": -0.0}
{"loss": -0.0}
{"loss": -0.0}
{"loss": -0.0}
{"loss": -0.0}
{"loss": -0.0}
{"loss": -0.0}
{"gpu_time": 26.91584014892578}
{"cpu_dispatch_time": 27.7307125}
{"cpu_walltime": 27.782303499999998}
{"#stdout": "GPU Time:             26.916 milliseconds"}
{"#stdout": "\n"}
{"#stdout": "CPU Dispatch Time:    27.731 milliseconds"}
{"#stdout": "\n"}
{"#stdout": "CPU Total Wall Time:  27.782 milliseconds"}
{"#stdout": "\n"}
{"#end": 1676344321.7381873, "#return_code": 0}
```

### pytorch_stargan

is it normal to have a bunch of `Start training...`?

```json
{"#start": 1677121533.0944624}
{"#config": {"dirs": {"code": "/network/scratch/o/ortizgas/public/bench_reqs/code/pytorch_stargan", "venv": "/network/scratch/o/ortizgas/public/bench_reqs/venv/pytorch_stargan", "data": "/network/scratch/o/ortizgas/public/bench_reqs/data", "runs": "/network/scratch/o/ortizgas/public/bench_reqs/runs", "base": "/network/scratch/o/ortizgas/public/bench_reqs"}, "venv": {"type": "virtualenv"}, "plan": {"method": "njobs", "n": 1}, "voir": {"--forward": ["#stdout", "#stderr", "#end", "loss", "reward", "compute_rate", "train_rate", "loading_rate", "descr", "progress", "total", "gpu_time", "cpu_dispatch_time", "cpu_walltime", "tflops"], "--compute-rate": true, "--train-rate": true, "--loading-rate": true}, "argv": {"-d": "cuda", "-m": "eager", "-t": "train"}, "config_base": "/home/satya-local/CODE/milabench_reqs/benchmarks/torchbench", "config_file": "/home/satya-local/CODE/milabench_reqs/benchmarks/torchbench/benchtest.yaml", "definition": "/home/satya-local/CODE/milabench_reqs/benchmarks/torchbench", "model": "pytorch_stargan", "name": "pytorch_stargan", "group": "pytorch_stargan", "tag": ["pytorch_stargan", "0"], "job-number": 0}}
{"#stdout": "Running train method from pytorch_stargan on cuda in eager mode."}
{"#stdout": "\n"}
{"#stdout": "Start training..."}
{"#stdout": "\n"}
{"#stderr": "/network/scratch/o/ortizgas/public/bench_reqs/venv/pytorch_stargan/lib/python3.9/site-packages/torch/nn/_reduction.py:42: UserWarning: size_average and reduce args will be deprecated, please use reduction='sum' instead.\n  warnings.warn(warning.format(ret))\n"}
{"#stdout": "Start training..."}
{"#stdout": "\n"}
{"#stdout": "Start training..."}
{"#stdout": "\n"}
{"#stdout": "Start training..."}
{"#stdout": "\n"}
{"#stdout": "Start training..."}
{"#stdout": "\n"}
{"#stdout": "Start training..."}
{"#stdout": "\n"}
{"#stdout": "Start training..."}
{"#stdout": "\n"}
{"#stdout": "Start training..."}
{"#stdout": "\n"}
{"#stdout": "Start training..."}
{"#stdout": "\n"}
{"#stdout": "Start training..."}
{"#stdout": "\n"}
{"#stdout": "Start training..."}
{"#stdout": "\n"}
{"train_rate": 269.85220498316806}
{"#stdout": "Start training..."}
{"#stdout": "\n"}
{"#stdout": "Start training..."}
{"#stdout": "\n"}
{"gpu_time": 58.62977600097656}
{"cpu_dispatch_time": 58.6149145}
{"cpu_walltime": 58.653855}
{"#stdout": "GPU Time:             58.630 milliseconds"}
{"#stdout": "\n"}
{"#stdout": "CPU Dispatch Time:    58.615 milliseconds"}
{"#stdout": "\n"}
{"#stdout": "CPU Total Wall Time:  58.654 milliseconds"}
{"#stdout": "\n"}
{"loss": -20.428274154663086}
{"#end": 1677121555.4614973, "#return_code": 0}
```

### squeezenet1_1

is it normal that the bench ends with a `squeezenet1_1`?

```json
{"#start": 1677121598.5013704}
{"#config": {"dirs": {"code": "/network/scratch/o/ortizgas/public/bench_reqs/code/squeezenet1_1", "venv": "/network/scratch/o/ortizgas/public/bench_reqs/venv/squeezenet1_1", "data": "/network/scratch/o/ortizgas/public/bench_reqs/data", "runs": "/network/scratch/o/ortizgas/public/bench_reqs/runs", "base": "/network/scratch/o/ortizgas/public/bench_reqs"}, "venv": {"type": "virtualenv"}, "plan": {"method": "njobs", "n": 1}, "voir": {"--forward": ["#stdout", "#stderr", "#end", "loss", "reward", "compute_rate", "train_rate", "loading_rate", "descr", "progress", "total", "gpu_time", "cpu_dispatch_time", "cpu_walltime", "tflops"], "--compute-rate": true, "--train-rate": true, "--loading-rate": true}, "argv": {"-d": "cuda", "-m": "eager", "-t": "train"}, "config_base": "/home/satya-local/CODE/milabench_reqs/benchmarks/torchbench", "config_file": "/home/satya-local/CODE/milabench_reqs/benchmarks/torchbench/benchtest.yaml", "definition": "/home/satya-local/CODE/milabench_reqs/benchmarks/torchbench", "model": "squeezenet1_1", "name": "squeezenet1_1", "group": "squeezenet1_1", "tag": ["squeezenet1_1", "0"], "job-number": 0}}
{"#stdout": "Running train method from squeezenet1_1 on cuda in eager mode."}
{"#stdout": "\n"}
{"#stderr": "/network/scratch/o/ortizgas/public/bench_reqs/venv/squeezenet1_1/lib/python3.9/site-packages/torchvision/models/_utils.py:208: UserWarning: The parameter 'pretrained' is deprecated since 0.13 and will be removed in 0.15, please use 'weights' instead.\n  warnings.warn(\n"}
{"#stderr": "/network/scratch/o/ortizgas/public/bench_reqs/venv/squeezenet1_1/lib/python3.9/site-packages/torchvision/models/_utils.py:223: UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and will be removed in 0.15. The current behavior is equivalent to passing `weights=SqueezeNet1_1_Weights.IMAGENET1K_V1`. You can also use `weights=SqueezeNet1_1_Weights.DEFAULT` to get the most up-to-date weights.\n  warnings.warn(msg)\n"}
{"gpu_time": 259.6874237060547}
{"cpu_dispatch_time": 259.6668605}
{"cpu_walltime": 259.7271315}
{"#stdout": "GPU Time:            259.687 milliseconds"}
{"#stdout": "\n"}
{"#stdout": "CPU Dispatch Time:   259.667 milliseconds"}
{"#stdout": "\n"}
{"#stdout": "CPU Total Wall Time: 259.727 milliseconds"}
{"#stdout": "\n"}
{"#stderr": "An error occurred"}
{"#stderr": "\n"}
{"#stderr": "Traceback (most recent call last):\n"}
{"#stderr": ""}
{"#stderr": "  File \"/network/scratch/o/ortizgas/public/bench_reqs/venv/squeezenet1_1/lib/python3.9/site-packages/voir/phase.py\", line 191, in _step_one\n    next_phase = gen.send(ph.value)\n"}
{"#stderr": ""}
{"#stderr": "  File \"/network/scratch/o/ortizgas/public/bench_reqs/venv/squeezenet1_1/lib/python3.9/site-packages/voir/overseer.py\", line 20, in __call__\n    yield ov.phases.run_script(priority=0)\n"}
{"#stderr": ""}
{"#stderr": "  File \"/network/scratch/o/ortizgas/public/bench_reqs/venv/squeezenet1_1/lib/python3.9/site-packages/giving/gvn.py\", line 590, in __exit__\n    obs.on_completed()\n"}
{"#stderr": ""}
{"#stderr": "  File \"/network/scratch/o/ortizgas/public/bench_reqs/venv/squeezenet1_1/lib/python3.9/site-packages/reactivex/observer/autodetachobserver.py\", line 46, in on_completed\n    self._on_completed()\n"}
{"#stderr": ""}
{"#stderr": "  File \"/network/scratch/o/ortizgas/public/bench_reqs/venv/squeezenet1_1/lib/python3.9/site-packages/reactivex/operators/_lastordefault.py\", line 29, in on_completed\n    observer.on_error(SequenceContainsNoElementsError())\n"}
{"#stderr": ""}
{"#stderr": "  File \"/network/scratch/o/ortizgas/public/bench_reqs/venv/squeezenet1_1/lib/python3.9/site-packages/reactivex/observer/autodetachobserver.py\", line 36, in on_error\n    self._on_error(error)\n"}
{"#stderr": ""}
{"#stderr": "  File \"/network/scratch/o/ortizgas/public/bench_reqs/venv/squeezenet1_1/lib/python3.9/site-packages/reactivex/observer/autodetachobserver.py\", line 36, in on_error\n    self._on_error(error)\n"}
{"#stderr": ""}
{"#stderr": "  File \"/network/scratch/o/ortizgas/public/bench_reqs/venv/squeezenet1_1/lib/python3.9/site-packages/reactivex/observer/autodetachobserver.py\", line 36, in on_error\n    self._on_error(error)\n"}
{"#stderr": ""}
{"#stderr": "  File \"/network/scratch/o/ortizgas/public/bench_reqs/venv/squeezenet1_1/lib/python3.9/site-packages/reactivex/internal/basic.py\", line 34, in default_error\n    raise err\n"}
{"#stderr": ""}
{"#stderr": "reactivex.internal.exceptions.SequenceContainsNoElementsError: Sequence contains no elements\n"}
{"#stderr": ""}
{"#end": 1677121620.5689182, "#return_code": 0}
```

**Comment by satyaog** [comment]
FYI

### soft_actor_critic

currently fails with

```json
{"#stderr": "AttributeError: 'PendulumEnv' object has no attribute 'seed'\n"}
```

### timm_efficientdet

latest error is

```json
{"#stderr": "TypeError: create_dataset() got an unexpected keyword argument 'custom_dataset_cfg'\n"}
```

File: benchmarks/torchbench/benchfile.py [hunk 3]
```diff
@@ -54,21 +55,25 @@ def install(self):
...
-        elif (code / f"requirements-{model_name}.txt").exists():
-            self.pip_install("-r", code / f"requirements-{model_name}.txt")
-        if (code / f"noinstallpy-{model_name}.txt").exists():
-            # We don't want to use the model's install.py, but we still want to
-            # run the global install.py, so we pass resnet18 which has an empty
-            # install.py.
-            self.python("install.py", "--models", "resnet18")
-        else:
+        with open(code / f"pre-post-{model_name}.out", "w") as _out, open(code / f"pre-post-{model_name}.err", "w") as _err:
+            self.execute("python3", "-m", "pip", "freeze", stdout=_out, stderr=_err)
```
Comment (breuleux) [inline benchmarks/torchbench/benchfile.py:63]: What does this do? Where are these files used? Or is this just for logging?

File: benchmarks/torchbench/benchfile.py [hunk 4]
```diff
@@ -54,21 +55,25 @@ def install(self):
...
+
+        if (code / "reqs" / model_name / "requirements-post.txt").exists():
+            self.pip_install("-r", code / "reqs" / model_name / "requirements-post.txt")
+
+        with open(code / f"pre-install-{model_name}.out", "w") as _out, open(code / f"pre-install-{model_name}.err", "w") as _err:
+            self.execute("python3", "-m", "pip", "freeze", stdout=_out, stderr=_err)
```
Comment (breuleux) [inline benchmarks/torchbench/benchfile.py:69]: Same question.

File: benchmarks/torchbench/reqs/setup.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
+# Fake setup.py to be used with piptools compile
```
Comment (breuleux) [inline benchmarks/torchbench/reqs/setup.py:1]: Why is this needed? I don't fully understand what's going on in this file.

File: benchmarks/torchbench/benchfile.py (hunk 3 above)
Comment (satyaog) [reply benchmarks/torchbench/benchfile.py:63]: I've used this to check what dependencies were not listed in an obvious way for each bench (during install.py) and add them to the requirements.in of the bench. It's not really need but useful for future benches or updates, should there be any. I can add a comment or remove those lines if you prefer

File: benchmarks/torchbench/reqs/setup.py (hunk 5 above)
Comment (satyaog) [reply benchmarks/torchbench/reqs/setup.py:1]: This is used in benchmarks/torchbench/reqs/pip-compile.sh, which is the script to build the reqs for all benches, and instead of writing a longer more complex bash script, I've offloaded some of the logic for fix some issues in python through those fake setup.py

**Review by breuleux — APPROVED** [review]
